README.md
[![Build Status](https://travis-ci.com/jehy/runtime-coverage.svg?branch=master)](https://travis-ci.com/jehy/runtime-coverage) [![npm version](https://badge.fury.io/js/runtime-coverage.svg)](https://badge.fury.io/js/runtime-coverage)
[![dependencies Status](https://david-dm.org/jehy/runtime-coverage/status.svg)](https://david-dm.org/jehy/runtime-coverage)
[![devDependencies Status](https://david-dm.org/jehy/runtime-coverage/dev-status.svg)](https://david-dm.org/jehy/runtime-coverage?type=dev)
[![Known Vulnerabilities](https://snyk.io/test/github/jehy/runtime-coverage/badge.svg)](https://snyk.io/test/github/jehy/runtime-coverage)
[![Coverage Status](https://coveralls.io/repos/github/jehy/runtime-coverage/badge.svg?branch=master)](https://coveralls.io/github/jehy/runtime-coverage?branch=master)

# Runtime coverage

Enable coverage any time after service startup, gather coverage and disable it!

Useful for integration tests and checking for dead code branches.

## Usage

### Install

```bash

npm install runtime-coverage

```

### Start coverage

```js

const runtimeCoverage = require('runtime-coverage');
await runtimeCoverage.startCoverage();

```

### Gather and output coverage

```js

const options = {
  reporters: ['text'],
  return: true,
  all: true,
  exclude: ['**/node_modules/**'],
  };
const res = await runtimeCoverage.getCoverage(options);
console.log(res.text);

```

You can also get coverage in any format of istanbul reporters, for example cobertura.

### Options

Options for `getCoverage`:

 * `{Object} options` options for getting coverage
 * `{Array} [options.exclude]` exclude those files in coverage, [micromatch](https://github.com/micromatch/micromatch) array, default `**/node_modules/**`
 * `{string} [options.rootDir]` starting directory for files that need coverage info, default `process.env.PWD`
 * `{boolean} [options.all]` include files which were not used in coverage data, default `false`
 * `{boolean} [options.deleteCoverage]` delete coverage directory after output, default `true`
 * `{string} [options.coverageDirectory]` Directory for storing coverage, defaults to temporary directory
 * `{boolean} [options.return]` return coverage data, default `true`.
 * `{boolean} [options.stream]` return coverage data as a stream, useful for huge coverage data,  default `false`
 * `{boolean} [options.streamTimeout]` destroy stream if not used after this timeout,  default `60*1000` millis
 * `{Array} [options.reporters]` Array of reporters to use, default "text", see all possible [here](https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-reports/lib).

if `return` option is true, result will be an object with coverage data, keys are reporters's
names. For example, for `text` reporter result will be in `res.text`, for `cobertura` it
will be `res.cobertura` and so on.

if `stream` option is true, object will have same structure, but instead of data there
will be a stream, which will be destroyed after `streamTimeout` if not used.

### Sample

Short express sample [here](https://github.com/jehy/runtime-coverage-sample).

### Beware

After first module call, in subsequent calls V8 only reports partial coverage for called
functions, and it leads to 100% module coverage. To avoid it, either:
* Call all your code after `startCoverage()` call
* Use option `forceLineMode` -  it can only help with line coverage and is approximate because
can't know about comments, empty lines and statements - but it's safe
* Use option `forceReload` (used by default) - it will try to safely reload and run your file. But
thinks may break, kitten may die, and devil may be summoned - I warned you!

V8 coverage is still experimental, so of cause some thing are pretty clumsy.
I'm still hoping to make a perfect runtime coverage - may be with help of istanbul metadata. Wanna help?

### Debug

If you wanna debug library work, launch your project with env variable
`DEBUG=runtime-coverage:* `


CHANGELOG.MD
# Changelog

Project is in experimental phase, so until version `1.0` there may be breaking changes everywhere.

### 0.0.14

* Empty coverage is now calculated in separate process to avoid any harm to running service;
* Result keys are now same as reporter name, not reporter default output file.


__tests__/test_passing.js
'use strict';

const os = require('os');
const assert = require('assert');
const fs = require('fs-extra');
const path = require('path');
const sinon = require('sinon');
const runtimeCoverage = require('../src');
const {add} = require('./test-lib/used');
const fixReport = require('../src/fixReport');

const REWRITE_SNAPSHOTS = false;

function createTempFile(data) {
  const tmpDir = fs.mkdtempSync(`${os.tmpdir()}${path.sep}`);
  const tmpFileName = `${tmpDir}${path.sep}tmp.txt`;
  fs.writeFileSync(tmpFileName, data);
  return {tmpDir, tmpFileName};
}

function compareSnapshot(testName, res) {
  const normalized = res.text.replace(/timestamp=".*"/g, 'timestamp=""').split(process.env.PWD).join('');
  // eslint-disable-next-line no-console
  console.log(normalized);
  const fileName = path.join(__dirname, `__snapshots__/${testName}.txt`);
  // for development
  if (REWRITE_SNAPSHOTS || !fs.existsSync(fileName)) {
    fs.writeFileSync(fileName, normalized);
  }
  const v8FileName = path.join(__dirname, `__snapshots__/${testName}.v8.json`);
  fs.writeFileSync(v8FileName, JSON.stringify(res.v8, null, 2));
  const expected = fs.readFileSync(fileName, 'utf8');
  assert.equal(normalized, expected);
}

describe('check coverage', ()=>{
  const standardExclude = ['**/index.js', '**/node_modules/**', '**/__tests__/test_*.js', '**/src/**'];
  const sandbox = sinon.createSandbox();
  afterEach(() => {
    sandbox.restore();
  });
  it('should include non used library when options.all is true', async ()=>{
    await runtimeCoverage.startCoverage();
    const testFunctionResult = add(1, 2);
    assert.equal(testFunctionResult, 3);
    const options = {
      reporters: ['text', 'v8'],
      return: true,
      all: true,
      exclude: standardExclude,
    };
    const res = await runtimeCoverage.getCoverage(options);
    compareSnapshot('allTrue', res);
  });

  it('sadly fails coverage on subsequent calls...', async () => {
    await runtimeCoverage.startCoverage();
    // eslint-disable-next-line global-require
    const testFunctionResult = add(1, 3);
    assert.equal(testFunctionResult, 4);
    const options = {
      reporters: ['text', 'v8'],
      return: true,
      all: true,
      exclude: standardExclude,
      forceReload: false,
    };
    const res = await runtimeCoverage.getCoverage(options);
    compareSnapshot('FAIL', res);
  });

  it('Does not fail if forceReload', async () => {
    await runtimeCoverage.startCoverage();
    const testFunctionResult = add(1, 3);
    assert.equal(testFunctionResult, 4);
    const options = {
      reporters: ['text', 'v8'],
      return: true,
      all: true,
      exclude: standardExclude,
    };
    const res = await runtimeCoverage.getCoverage(options);
    compareSnapshot('allTrueReloaded', res);
  });

  it('should not include non used library when options.all is false', async ()=>{
    await runtimeCoverage.startCoverage();
    const testFunctionResult = add(3, 2);
    assert.equal(testFunctionResult, 5);
    const options = {
      reporters: ['text', 'v8'],
      return: true,
      exclude: standardExclude,
      forceReload: true,
    };
    const res = await runtimeCoverage.getCoverage(options);
    compareSnapshot('allFalse', res);
  });

  it('should use rootDir for filtering coverage data', async ()=>{
    await runtimeCoverage.startCoverage();
    const testFunctionResult = add(1, 2);
    assert.equal(testFunctionResult, 3);
    const options = {
      reporters: ['text', 'v8'],
      return: true,
      all: true,
      rootDir: path.join(__dirname, 'test-lib'),
    };
    const res = await runtimeCoverage.getCoverage(options);
    compareSnapshot('rootDir', res);
  });

  it('should count lines approximately fine with forceLineMode even if module was already called', async ()=>{
    await runtimeCoverage.startCoverage();
    const testFunctionResult = add(1, 2);
    assert.equal(testFunctionResult, 3);
    const options = {
      reporters: ['text', 'v8'],
      return: true,
      all: true,
      exclude: standardExclude,
      forceLineMode: true,
      forceReload: false,
    };
    const res = await runtimeCoverage.getCoverage(options);
    compareSnapshot('forceLineMode', res);
  });

  it('should work even if file with calls is covered', async ()=>{
    await runtimeCoverage.startCoverage();
    // eslint-disable-next-line global-require
    require('./test-lib/complex');
    const options = {
      reporters: ['text', 'v8'],
      return: true,
      exclude: standardExclude,
    };
    const res = await runtimeCoverage.getCoverage(options);
    compareSnapshot('complex', res);
  });

  it('should work with cobertura, fix it, give a stream and self destroy extra stream', async ()=>{
    const spy = sandbox.spy(fixReport, 'fixCoberturaReport');
    await runtimeCoverage.startCoverage();
    const options = {
      reporters: ['cobertura', 'json'],
      return: true,
      stream: true,
      streamTimeout: 2 * 1000,
      exclude: standardExclude,
    };
    const testFunctionResult = add(1, 2);
    assert.equal(testFunctionResult, 3);
    const res = await runtimeCoverage.getCoverage(options);
    const data = [];
    const stream = res.cobertura;
    stream.on('data', chunk => data.push(chunk));
    await new Promise((resolve, reject)=>{
      stream.on('end', () => resolve());
      stream.on('error', err => reject(err));
    });
    assert(spy.called);
    compareSnapshot('stream', {text: data.join(), v8: res.v8});
  });

  it('should really fix cobertura report', async ()=>{
    const dataToFix = '<some tag>some data</some tag><computed>lalala</computed>';
    const {tmpDir, tmpFileName} = createTempFile(dataToFix);
    await fixReport.fixCoberturaReport(tmpFileName);
    const newData = await fs.readFile(tmpFileName);
    await fs.remove(tmpDir);
    assert.equal(newData, '<some tag>some data</some tag>&lt;computed&gt;lalala</computed>');
  });
});


__tests__/test-lib/used.js
'use strict';


function add(a, b) {
  if (1 === 2) {
    // eslint-disable-next-line no-console
    console.log('what a day!');
  }
  return a + b;
}

// eslint-disable-next-line no-unused-vars
function theWorld() {
  return 42;
}

module.exports = {add};


__tests__/test-lib/complex.js
'use strict';

const fs = require('fs');
const path = require('path');

// eslint-disable-next-line no-unused-vars
function lol() {
  // eslint-disable-next-line no-console
  console.log(null);
}

// eslint-disable-next-line no-console
console.log(fs.readFileSync(path.join(__dirname, 'used.js'), 'utf8').split('\n')[0]);


src/utils.js
'use strict';

const Debug = require('debug');
const micromatch = require('micromatch');
const path = require('path');

const debug = {
  reporters: Debug('runtime-coverage:debug-reporters'),
  emptyCov: Debug('runtime-coverage:empty-coverage'),
  mergeCov: Debug('runtime-coverage:merge-full-coverage'),
  getCov: Debug('runtime-coverage:get-coverage'),
  startCov: Debug('runtime-coverage:get-coverage'),
};

function match(itemPath, excludeArray) {
  // HACK for https://github.com/paulmillr/chokidar/issues/577
  const basename = path.basename(itemPath);
  if (basename[0] === '.') {
    return micromatch.isMatch(itemPath.replace(basename, basename.substr(1)), excludeArray);
  }
  return micromatch.isMatch(itemPath, excludeArray);
}

function shouldCover(fileName, options) {
  return fileName.startsWith(options.rootDir) && fileName.endsWith('.js') && !match(fileName, options.exclude);
}

module.exports = {debug, shouldCover};


src/fixReport.js
'use strict';

const { spawn } = require('child_process');
const Promise = require('bluebird');

async function fixCoberturaReport(fileName) {
  // workaround for https://github.com/istanbuljs/istanbuljs/issues/527
  // sed -i 's/<computed>/\&lt;computed\&gt;/g' ./cobertura-coverage.xml
  return new Promise((resolve, reject) => {
    const stdout = [];
    const stderr = [];
    // eslint-disable-next-line no-useless-escape
    const ps = spawn('sed', ['-i', 's/<computed>/\\&lt;computed\\&gt;/g', fileName], {});
    ps.stdout.on('data', (newData) => {
      stdout.push(newData);
    });

    ps.stderr.on('data', (newData) => {
      stderr.push(newData);
    });

    // eslint-disable-next-line no-unused-vars
    ps.on('close', (code) => {
      const data = { stderr: stderr.join('\n'), stdout: stdout.join('\n'), code};
      if (code === 0) {
        resolve(data);
        return;
      }
      reject(data);
    });
  });
}

module.exports = {fixCoberturaReport};


src/v8-coverage.js
'use strict';

const {CoverageMap} = require('istanbul-lib-coverage/lib/coverage-map');

// Code for merging reports taken from https://github.com/Eywek/v8-coverage

/**
 * Merge previous map with one report data
 * @param {CoverageMap} map
 * @param {Object} reportData
 */
function mergeReportData(map, reportData) {
  Object.keys(reportData.branchMap).forEach((k) => {
    if (!map.data.branchMap[k]) {
      map.data.branchMap[k] = reportData.branchMap[k];
    }
  });
  Object.keys(reportData.fnMap).forEach((k) => {
    if (!map.data.fnMap[k]) {
      map.data.fnMap[k] = reportData.fnMap[k];
    }
  });
  Object.keys(reportData.statementMap).forEach((k) => {
    if (!map.data.statementMap[k]) {
      map.data.statementMap[k] = reportData.statementMap[k];
    }
  });
  Object.keys(reportData.s).forEach((k) => {
    map.data.s[k] += reportData.s[k];
  });
  Object.keys(reportData.f).forEach((k) => {
    map.data.f[k] += reportData.f[k];
  });
  Object.keys(reportData.b).forEach((k) => {
    const retArray = map.data.b[k];
    const secondArray = reportData.b[k];
    if (!retArray) {
      map.data.b[k] = secondArray;
      return;
    }
    for (let i = 0; i < retArray.length; i += 1) {
      retArray[i] += secondArray[i];
    }
  });
}


/**
 * Merge previous map with one report
 * @param {CoverageMap} map
 * @param {Object} report
 */
function mergeReport(map, report) {
  let sourceMap;
  if (report instanceof CoverageMap) {
    sourceMap = report;
  } else {
    sourceMap = new CoverageMap(report);
  }
  Object.keys(sourceMap.data).forEach((k) => {
    const fc = sourceMap.data[k];
    if (map.data[k]) {
      mergeReportData(map.data[k], fc);
    } else {
      map.data[k] = fc;
    }
  });
}
/**
 * Merge previous map with reports
 * @param {CoverageMap} map
 * @param {Array} reports
 */
function mergeMap(map, reports) {
  reports.forEach((report) => {
    mergeReport(map, report);
  });
  return map;
}

module.exports = {mergeMap};


src/fixCoberturaReport.js
'use strict';

const { spawn } = require('child_process');
const Promise = require('bluebird');

async function fixCoberturaReport(fileName) {
  // workaround for https://github.com/istanbuljs/istanbuljs/issues/527
  // sed -i 's/<computed>/\&lt;computed\&gt;/g' ./cobertura-coverage.xml
  return new Promise((resolve, reject) => {
    const stdout = [];
    const stderr = [];
    // eslint-disable-next-line no-useless-escape
    const ps = spawn('sed', ['-i', 's/<computed>/\\&lt;computed\\&gt;/g', fileName], {});
    ps.stdout.on('data', (newData) => {
      stdout.push(newData);
    });

    ps.stderr.on('data', (newData) => {
      stderr.push(newData);
    });

    // eslint-disable-next-line no-unused-vars
    ps.on('close', (code) => {
      const data = { stderr: stderr.join('\n'), stdout: stdout.join('\n'), code};
      if (code === 0) {
        resolve(data);
        return;
      }
      reject(data);
    });
  });
}

module.exports = fixCoberturaReport;


src/getEmptyV8Coverage.js
'use strict';

const Module = require('module');
const path = require('path');
const { CoverageInstrumenter } = require('collect-v8-coverage');
const { fileURLToPath } = require('url');
const Promise = require('bluebird');

const {debug, shouldCover} = require('./utils');

const infiniteHandler = {
  get(obj, prop) {
    debug.emptyCov(`tried to access prop ${prop}`);
    return this;
  },
};

const infiniteProxy = new Proxy({}, infiniteHandler);

async function getEmptyV8Coverage(files, options) {
  const v8CoverageInstrumenter2 = new CoverageInstrumenter();
  await v8CoverageInstrumenter2.startInstrumenting();
  const originalRequire = Module.prototype.require;

  Module.prototype.require = (filename) => {
    debug.emptyCov(`${path.basename(filename)} Attempted to require: ${filename}`);
    return infiniteProxy;
  };
  files.forEach((file) => {
    const cache = require.cache[require.resolve(file)];
    require.cache[require.resolve(file)] = undefined;
    try {
      const tempModule = originalRequire(file);
      Object.values(tempModule)
        .forEach((someFunc)=>{
          // try to call all module funcs for maximum detailed empty coverage
          try {
            if (typeof someFunc !== 'function') {
              return;
            }
            someFunc();
          } catch (err) {
            debug.emptyCov(`func call failed: ${err}`);
          }
        });
    } catch (err) {
      debug.emptyCov(`Require failed: ${err}`);
    } finally {
      require.cache[require.resolve(file)] = cache;
    }
  });
  Module.prototype.require = originalRequire;
  const coverage = await v8CoverageInstrumenter2.stopInstrumenting();

  return coverage
    .filter(res => res.url.startsWith('file://'))
    .map((res)=>{
      return { ...res, url: fileURLToPath(res.url)};
    })
    .filter(res => res && shouldCover(res.url, options))
    .map((res) => {

      const functions = res.functions.map((func)=>{
        if (func.functionName === '') {
          return func;
        }
        return {
          ...func,
          ranges: func.ranges.map((range)=>{
            return {...range, count: 0};
          }),
        };
      });

      return { ...res, functions};
    });
}


process.on('message', async (message) => {
  try {
    const coverageResult = await getEmptyV8Coverage(message.files, message.options);
    const res = await Promise.promisify(process.send, {context: process})(coverageResult, undefined, {});
    if (res !== true) {
      process.exit(1);
    }
  } catch (err) {
    // eslint-disable-next-line no-console
    await Promise.promisify(process.send, {context: process})(err, undefined, {}).catch((err2)=>{ console.log(err2); });
    process.exit(1);
  }
  process.exit(0);
});


src/index.js
'use strict';

const { CoverageInstrumenter } = require('collect-v8-coverage');
const v8ToIstanbul = require('v8-to-istanbul');
const libCoverage = require('istanbul-lib-coverage');
const libReport = require('istanbul-lib-report');
const istanbulReports = require('istanbul-reports');
const { fileURLToPath } = require('url');
const path = require('path');
const fs = require('fs-extra');
const os = require('os');
const { fork } = require('child_process');
const Promise = require('bluebird');
const klawSync = require('klaw-sync');
const semver = require('semver');

const {mergeMap} = require('./v8-coverage');
const {debug, shouldCover} = require('./utils');
const fixReport = require('./fixReport');

let v8CoverageInstrumenter;

async function getEmptyV8Coverage(files, options) {
  const getEmptyV8CoverageProcess = fork(path.join(__dirname, 'getEmptyV8Coverage'), [], {env: {}, execArgv: []});
  getEmptyV8CoverageProcess.send({ files, options });
  let replyReceived = false;
  try {
    // we need to await here, otherwise finally won't work
    // eslint-disable-next-line sonarjs/prefer-immediate-return
    const result = await new Promise((resolve, reject) => {
      getEmptyV8CoverageProcess
        .on('message', (message) => {
          replyReceived = true;
          if (message instanceof Error) {
            reject(message);
            return;
          }
          resolve(message);
        })
        .on('error', err => reject(err))
        .on('exit', (code, signal) => reject(new Error(`process closed before data was sent! code: ${code} signal: ${signal}`)));
    });
    return result;
  }  finally {
    if (!replyReceived) {
      debug.emptyCov('Reply not received, killing empty coverage process');
      getEmptyV8CoverageProcess.kill('SIGTERM');
    }
  }
}


function getAllProjectFiles(options) {
  const filterFunc = file=>shouldCover(file.path, options);
  const files = klawSync(options.rootDir, {filter: filterFunc, traverseAll: true, nodir: true});
  return files.map(file => file.path);
}

function normalizeOptions(options) {
  if (!options) {
    throw new Error('Options not passed!');
  }
  options.exclude = options.exclude || ['**/node_modules/**'];
  options.rootDir = path.normalize(options.rootDir || process.env.PWD);
  options.all = options.all || false;
  options.forceLineMode = options.forceLineMode || false;
  options.streamTimeout = options.streamTimeout || 60 * 1000;
  options.deleteCoverage = options.deleteCoverage || options.deleteCoverage === undefined;
  options.coverageDirectory = path.normalize(options.coverageDirectory || fs.mkdtempSync(`${os.tmpdir()}${path.sep}`));
  options.return = options.return || options.return === undefined;
  options.forceReload = options.forceReload || options.forceReload === undefined;
  if (!options.reporters || !options.reporters.length) {
    options.reporters = ['text'];
  }
  if (options.coverageDirectory === options.rootDir) {
    throw new Error(`Coverage directory ${options.coverageDirectory} should be different from project directory!`);
  }
  if (options.coverageDirectory.split(path.sep).length < 3) {
    throw new Error(`Suspicious coverage dir ${options.coverageDirectory}`);
  }
}

function createEmptyCoverageBlock(file) {
  const fileSize = fs.lstatSync(file).size;
  return {
    functionName: '',
    ranges: [{
      startOffset: 0,
      endOffset: fileSize,
      count: 0,
    }],
    isBlockCoverage: true,
  };
}

async function runReporters(options, map, coverageData) {
  const context = libReport.createContext({
    dir: options.coverageDirectory,
    coverageMap: map,
  });

  options.reporters.forEach((reporter) => {
    if (reporter === 'v8') {
      return;// we will deal with it later
    }
    const reportOptions = {
      file: reporter,
    };
    const report = istanbulReports.create(reporter, reportOptions);
    report.execute(context);
  });
  let res = true;
  const allStreamsClosed = [];
  if (options.return) {
    const files = fs.readdirSync(options.coverageDirectory).filter(f=>f !== '.' && f !== '..');
    res = await Promise.reduce(files,  async (data, filename)=>{
      const filePath = path.join(options.coverageDirectory, filename);
      if (fs.lstatSync(filePath).isDirectory()) {
        // ignore dirs
        return data;
      }
      debug.reporters(`got filename ${filename}`);
      if (filename.includes('cobertura')) {
        await fixReport.fixCoberturaReport(filePath);
        debug.reporters(`fixed cobertura file ${filename}`);
        // data[filename] = data[filename].replace(/<computed>/g, '&lt;computed&gt;');
      }
      if (options.stream) {
        const stream = fs.createReadStream(filePath, {encoding: 'utf8', emitClose: true});
        data[filename] = stream;
        const streamClosed = new Promise((resolve, reject)=>{
          stream.once('error', err=>reject(err));
          stream.once('close', ()=>resolve());
        });
        allStreamsClosed.push(streamClosed);
        streamClosed.timeout(options.streamTimeout).catch(Promise.TimeoutError, ()=>{
          debug.reporters(`No one uses stream ${filePath}, destroying it...`);
          stream.destroy();
        }).catch((err)=>{
          debug.reporters('failed to destroy stream', err);
        });
      } else {
        data[filename] = await fs.readFile(filePath, 'utf8');
      }
      return data;
    }, {});
  }
  if (options.deleteCoverage) {
    Promise.all(allStreamsClosed).catch((err)=>{
      debug.reporters('Failed read coverage data in stream', err);
    }).finally(async ()=>{
      try {
        await fs.remove(options.coverageDirectory);
        debug.reporters('coverage directory removed');
      } catch (err) {
        debug.reporters('Failed to remove coverage directory', err);
      }
    });
  }
  if (options.reporters.includes('v8')) {
    // for debug only, don't bother with a stream, either way it's already in memory
    res.v8 = coverageData;
  }
  return res;
}

function MergeFullCoverage(coverageData, emptyCoverage) {
  coverageData.forEach((report)=>{
    const emptyReport = emptyCoverage.find((rep)=>{
      return rep.url === report.url;
    });
    if (!emptyReport) {
      debug.mergeCov(`No empty report for ${report.url}, smth went wrong?`);
      return;
    }
    Object.values(emptyReport.functions).forEach((missingFunc)=>{
      const coveredFunc = report.functions.find(el=>el.functionName === missingFunc.functionName);
      if (!coveredFunc) {
        debug.mergeCov(`Adding func "${missingFunc.functionName}"`);
        if (missingFunc.functionName === '') {
          report.functions.unshift(missingFunc);
        } else {
          report.functions.push(missingFunc);
        }
      }  else {
        debug.mergeCov(`checking ranges for adding func "${missingFunc.functionName}":`);
        debug.mergeCov(`missing: ${JSON.stringify(missingFunc.ranges)}`);
        debug.mergeCov(`existing func ranges: ${JSON.stringify(coveredFunc.ranges)}`);
        // check missing ranges
        const coveredRanges = Object.values(coveredFunc.ranges).map(el=>`${el.startOffset}-${el.endOffset}`);
        Object.values(missingFunc.ranges).forEach((range)=>{
          debug.mergeCov(`checking if range hash ${Object.values(range).join('-')} is in ${JSON.stringify(coveredRanges)}`);
          if (!coveredRanges.includes(`${range.startOffset}-${range.endOffset}`)) {
            coveredFunc.ranges.push(range);
            debug.mergeCov(`Pushed range ${range.startOffset}-${range.endOffset}`);
          }
        });
      }
    });
  });
}


/**
 * @param {Object}options options for getting coverage
 * @param {Array} [options.exclude] exclude those files in coverage,
 * micromatch (https://github.com/micromatch/micromatch) array, default node modules
 * @param {string} [options.rootDir] starting directory for files that need coverage info, default process.env.PWD
 * @param {boolean} [options.all] include files which were not used in coverage data, default false
 * @param {boolean} [options.deleteCoverage] delete coverage directory after output, default true
 * @param {boolean} [options.forceLineMode] force per line coverage, it works with subsequent function calls
 * @param {boolean} [options.forceReload] reload modules to get full coverage data
 * @param {string} [options.coverageDirectory] Directory for storing coverage, defaults to temporary directory
 * @param {boolean} [options.return] return coverage data
 * @param {boolean} [options.stream] return coverage data as streams
 * @param {boolean} [options.streamTimeout] destroy stream if not used during timeout
 * @param {Array} [options.reporters] Array of reporters to use, default "text"
 * https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-reports/lib
 *
 */
async function getCoverage(options) {
  debug.getCov(`getCoverage called. Options: ${JSON.stringify(options)}`);
  normalizeOptions(options);
  debug.getCov(`normalized options: ${JSON.stringify(options)}`);
  if (!v8CoverageInstrumenter) {
    throw new Error('You need to start coverage first!');
  }
  const v8CoverageResult = await v8CoverageInstrumenter.stopInstrumenting();
  debug.getCov('stopped instrumenting');
  v8CoverageInstrumenter = null;

  const coverageData = v8CoverageResult
    .filter(res => res.url.startsWith('file://'))
    .map(res => ({ ...res, url: fileURLToPath(res.url) }))
    .filter(res => shouldCover(res.url, options));
  v8CoverageResult.length = 0; // we don't need it any more

  const coveredFiles = coverageData.map(data=>data.url);
  if (options.forceReload) {
    // get empty coverage data for merging later
    const emptyCoverage = await getEmptyV8Coverage(coveredFiles, options);
    MergeFullCoverage(coverageData, emptyCoverage);
  }
  // process actual coverage together with empty from prev step(if options.forceReload)
  const reportsListCovered = await Promise.map(coverageData, async (report) => {
    const converter = v8ToIstanbul(report.url);
    await converter.load(); // this is required due to the async source-map dependency.
    if (options.forceLineMode) {
      const emptyFunction = report.functions.find(func => func.functionName === '');
      if (!emptyFunction) {
        report.functions.unshift(createEmptyCoverageBlock(report.url));
      }
    }
    converter.applyCoverage(report.functions);
    return converter.toIstanbul();
  }, {concurrency: 1});
  if (!options.reporters.includes('v8')) {
    // drop it, not needed any more
    coverageData.length = 0;
  }
  let reportsList;
  // generate dummy empty coverage for non required files
  if (options.all) {
    const emptyReports = getAllProjectFiles(options)
      .filter(file => !coveredFiles.includes(file));
    const reportsListEmpty = await Promise.map(emptyReports, async (file) => {
      const converter = v8ToIstanbul(file);
      try {
        // Since some packages are build and their source is removed, they
        // can leave unsatisfied dependants. So just ignore it.
        await converter.load();
      } catch (err) {
        debug.getCov('Error on converter.load() for all files', JSON.stringify(err));
        return false;
      }
      converter.applyCoverage([createEmptyCoverageBlock(file)]); // apply empty coverage
      return converter.toIstanbul();
    }, {concurrency: 1});
    reportsList = reportsListCovered.concat(reportsListEmpty.filter(Boolean));
  } else {
    reportsList = reportsListCovered;
  }

  const map = mergeMap(libCoverage.createCoverageMap({}), reportsList);
  // we don't need them any more. V8 should clean it up either way, but just to be sure...
  reportsListCovered.length = 0;
  reportsList.length = 0;
  return runReporters(options, map, coverageData);
}


async function startCoverage() {
  if (!semver.satisfies(process.version, '>= 10.12.0')) {
    throw new Error(`Node version ${process.version} does not have coverage information!
    Please use node >= 10.12.0 or babel coverage.`);
  }
  if (v8CoverageInstrumenter) {
    debug.startCov('v8CoverageInstrumenter already started, restarting it');
    await v8CoverageInstrumenter.stopInstrumenting();
    v8CoverageInstrumenter = null;
  }
  v8CoverageInstrumenter = new CoverageInstrumenter();
  await v8CoverageInstrumenter.startInstrumenting();
  debug.startCov('started instrumenting');
  return true;
}


module.exports = {
  startCoverage,
  getCoverage,
};
